sdks/go/node/core/containerruntime/docker/imagePuller.go

```diff
@@ -38,12 +40,41 @@ type _imagePuller struct {
 
 func (ip _imagePuller) Pull(
 	ctx context.Context,
-	containerID string,
-	imagePullCreds *model.Creds,
-	imageRef string,
+	containerCall *model.ContainerCall,
 	rootCallID string,
 	eventPublisher pubsub.EventPublisher,
 ) error {
+	imageRef := *containerCall.Image.Ref
+	imagePullCreds := containerCall.Image.PullCreds
+	containerID := containerCall.ContainerID
+
+	// Skip pulling for non-tagged images that already are present
+	// This reduces the chance of hitting docker rate limiting errors
+	// and speeds up execution.
+	ref, err := reference.ParseAnyReference(strings.ToLower(imageRef))
+	if err != nil {
+		return err
+	}
+	named, err := reference.ParseNormalizedNamed(ref.String())
+	if err != nil {
+		return err
+	}
+	if tagged, ok := named.(reference.Tagged); ok && tagged.Tag() != "latest" {
+		_, _, err := ip.dockerClient.ImageInspectWithRaw(ctx, imageRef)
+		if err == nil {
+			eventPublisher.Publish(model.Event{
+				Timestamp: time.Now().UTC(),
+				ContainerStdOutWrittenTo: &model.ContainerStdOutWrittenTo{
+					Data:        []byte(fmt.Sprintf("Skipping image pull: %s\n", imageRef)),
+					OpRef:       containerCall.OpPath,
+					ContainerID: containerCall.ContainerID,
+					RootCallID:  rootCallID,
+				},
+			})
+
+			return nil
+		}
+	}
```

Comment:
Suggestion: wrap this whole new code into function (`func checkImageExistsAndUpToDate(...)` or sth)